python/src/otel/otel_sdk/otel-instrument

```diff
@@ -1,44 +1,160 @@
+`opentelemetry-instrument` auto instrumentation script from the
+`opentelemetry-instrumentation` package.
+
+Additionally, this configuration the user is using packages derived from
+`opentelemetry-instrumentation` and `opentelemetry-sdk` in their implementation.
+
+Usage
+-----
+In the configuration of an AWS Lambda function with
+`otel-instrument` at the root level of a Lambda
+Layer:
+
+.. code::
+
+    AWS_LAMBDA_EXEC_WRAPPER = /opt/otel-instrument
+
+"""
+
+import os
 import sys
+from os import environ, execl
 
-# the path to the interpreter and all of the originally intended arguments
-args = sys.argv[1:]
+AWS_LAMBDA_FUNCTION_NAME = "AWS_LAMBDA_FUNCTION_NAME"
+PYTHONPATH = "PYTHONPATH"
+LAMBDA_RUNTIME_DIR = "LAMBDA_RUNTIME_DIR"
 
-# enable OTel wrapper
-environ["ORIG_HANDLER"] = environ.get("_HANDLER")
-environ["_HANDLER"] = "otel_wrapper.lambda_handler"
+# Update the python paths for packages with `sys.path` and `PYTHONPATH`
+
+# - Expect this `otel-instrument` file to be at the Lambda Layer root. Then, we
+#   know where the OpenTelemetry Python packages are and can add them to the
+#   PYTHONPATH.
+
+LAMBDA_LAYER_PKGS_DIR = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), "python")
+)
+
+# - Set Lambda Layer python packages in PYTHONPATH so `opentelemetry-instrument`
+#   script can find them (it needs to find `opentelemetry` find the auto
+#   instrumentation `run()` method)
+
+if PYTHONPATH not in environ:
+    environ[PYTHONPATH] = LAMBDA_LAYER_PKGS_DIR
+elif LAMBDA_LAYER_PKGS_DIR not in environ[PYTHONPATH]:
+    environ[PYTHONPATH] += os.pathsep + LAMBDA_LAYER_PKGS_DIR
+
+# - Set Lambda runtime python packages in PYTHONPATH so
+#   `opentelemetry-instrument` script can find them during auto instrumentation
+#   and instrument them.
+
+if PYTHONPATH not in environ:
+    environ[PYTHONPATH] = os.environ[LAMBDA_RUNTIME_DIR]
+if os.environ[LAMBDA_RUNTIME_DIR] not in environ[PYTHONPATH]:
+    environ[PYTHONPATH] += os.pathsep + os.environ[LAMBDA_RUNTIME_DIR]
+
+# Configure OpenTelemetry Python with environment variables
 
-# config default traces exporter if missing
-environ.setdefault("OTEL_TRACES_EXPORTER", "otlp_proto_grpc_span")
+# - Set Lambda Layer python packages in current python path so we can find them
+#   right away in this script
 
-# set service name
-if environ.get("OTEL_RESOURCE_ATTRIBUTES") is None:
-    environ["OTEL_RESOURCE_ATTRIBUTES"] = "service.name=%s" % (
-        environ.get("AWS_LAMBDA_FUNCTION_NAME")
+if LAMBDA_LAYER_PKGS_DIR not in sys.path:
+    sys.path.append(LAMBDA_LAYER_PKGS_DIR)
+
+from opentelemetry.environment_variables import (
+    OTEL_PROPAGATORS,
+    OTEL_TRACES_EXPORTER,
+    OTEL_PYTHON_DISABLED_INSTRUMENTATIONS,
+)
+from opentelemetry.sdk.environment_variables import OTEL_RESOURCE_ATTRIBUTES
+
+# - Set the default Trace Exporter
+
+environ.setdefault(OTEL_TRACES_EXPORTER, "otlp_proto_grpc_span")
+
+# - Set the service name
+
+service_name_resource_attribute = "service.name"
```

Comment:
This seems to not respect the user's service.name

Reply:
Actually just use OTEL_SERVICE_NAME instead, respecting the default if already set

https://github.com/open-telemetry/opentelemetry-python/blob/18b5cb046bebab0c70841c662268bb7980c2fd5e/opentelemetry-sdk/src/opentelemetry/sdk/environment_variables.py#L356